# Quantum error correction - bit-flip code quantum kata Workbook

**What is this workbook?** 
A workbook is a collection of problems, accompanied by solutions to them. The explanations focus on the logical steps required to solve a problem; they illustrate the concepts that need to be applied to come up with a solution to the problem, explaining the mathematical steps required.

Note that a workbook should not be the primary source of knowledge on the subject matter; it assumes that you've already read a tutorial or a textbook and that you are now seeking to improve your problem-solving skills. You should attempt solving the tasks of the respective kata first, and turn to the workbook only if stuck or for reinforcement. While a textbook emphasizes knowledge acquisition, a workbook emphasizes skill acquisition.

This workbook describes the solutions to the problems offered in the [QEC Bit Flip Code Kata](./QEC_BitFlipCode.ipynb). 
Since the tasks are offered as programming problems, the explanations also cover some elements of Q# that might be non-obvious for a novitiate.

**What you should know for this workbook**

You should be familiar with the following concepts and associated techniques **prior to** beginning work on the Basic Gates Quantum Kata.

1. 

You can also consult the [complete Quantum Katas learning path](https://github.com/microsoft/QuantumKatas#learning-path).

### Task 1. Parity Measurements

**Input:** three qubits (stored as an array of length 3) in an unknown basis state or in a superposition of basis states of the same parity.

**Output:** the parity of this state encoded as a value of `Result` type: `Zero` for parity 0 and `One` for parity 1.  The parity of basis state $| x_{0} x_{1} x_{2}\rangle$ is defined as $(x_{0} \oplus x_{1} \oplus x_{2})$.  After applying the operation the state of the qubits should not change. You can use exactly one measurement.
    
**Example:** $|000 \rangle$, $|101\rangle$ and $|011\rangle$ all have parity 0, while $|010\rangle$ and $|111\rangle$ have parity 1.

### Solution

Parity explanation
$$
|000\rangle \rightarrow (0 + 0 + 0) \text{ mod } 2 = 0 \\
|101\rangle \rightarrow (1 + 0 + 1) \text{ mod } 2 = 0 \\
|011\rangle \rightarrow (0 + 1 + 1) \text{ mod } 2 = 0 \\

|111\rangle \rightarrow (1 + 1 + 1) \text{ mod } 2 = 1 \\
|010\rangle \rightarrow (0 + 1 + 0) \text{ mod } 2 = 1 \\
$$

In [1]:
%kata T01_MeasureParity 

operation MeasureParity (register : Qubit[]) : Result {
    return Measure([PauliZ, PauliZ, PauliZ], register);
}

Testing on |000⟩ + |011⟩ with parity 0
Testing on |000⟩ + |101⟩ with parity 0
Testing on |000⟩ + |110⟩ with parity 0
Testing on |001⟩ + |010⟩ with parity 1
Testing on |001⟩ + |100⟩ with parity 1
Testing on |001⟩ + |111⟩ with parity 1
Testing on |010⟩ + |100⟩ with parity 1
Testing on |010⟩ + |111⟩ with parity 1
Testing on |011⟩ + |101⟩ with parity 0
Testing on |011⟩ + |110⟩ with parity 0
Testing on |100⟩ + |111⟩ with parity 1
Testing on |101⟩ + |110⟩ with parity 0


Success!

[Return to Task 1. of the QEC Bit Flip Code Kata.](./QEC_BitFlipCode.ipynb#Task-1.-Parity-Measurements)

### Task 2. Encoding Codewords

**Input**: three qubits in the state $| \psi \rangle \otimes |00\rangle$, where $|\psi\rangle = \alpha |0\rangle + \beta |1\rangle$ is the state of the first qubit, i.e., `register[0]`.

**Goal**:  create a state $|\bar{\psi}\rangle := \alpha |000\rangle + \beta |111\rangle$ on these qubits.   

### Solution

Explaination

In [2]:
%kata T02_Encode 

open Microsoft.Quantum.Arrays;
open Microsoft.Quantum.Intrinsic;
open Microsoft.Quantum.Canon;

operation Encode (register : Qubit[]) : Unit {
    ApplyToEachA(CNOT(Head(register), _), Rest(register));
}

Success!

[Return to Task 2. of the QEC Bit Flip Code Kata.](./QEC_BitFlipCode.ipynb#Task-2.-Encoding-Codewords)

### Task 3. Error Detection I

**Inputs:** three qubits that are either in the state $|\bar{\psi}\rangle = \alpha |000\rangle + \beta |111\rangle$ or in the state $X\mathbb{11} |\bar{\psi}\rangle = \alpha |100\rangle + \beta |011\rangle$.  

> Note that the second state is the first state with X applied to the first qubit, which corresponds to an X error happening on the first qubit.

**Output:** `Zero` if the input is $|\bar{\psi}\rangle$ (state without the error), `One` if the input is $X\mathbb{11} |\bar{\psi}\rangle$ (state with the error).  After applying the operation the state of the qubits should not change.

### Solution

Explaination

In [3]:
%kata T03_DetectErrorOnLeftQubit 

operation DetectErrorOnLeftQubit (register : Qubit[]) : Result {
    return Measure([PauliZ, PauliZ], register[0 .. 1]);
}

Success!

[Return to Task 3. of the QEC Bit Flip Code Kata.](./QEC_BitFlipCode.ipynb#Task-3.-Error-Detection-I)

### Task 4. Error Correction I

**Input:** three qubits that are either in the state $|\bar{\psi}\rangle = \alpha |000\rangle + \beta |111\rangle$ or in the state $X\mathbb{11} |\bar{\psi}\rangle = \alpha |100\rangle + \beta |011\rangle$.

**Goal:**  make sure that the qubits are returned to the state $|\bar{\psi}\rangle$  (i.e., determine whether an X error has occurred, and if so, fix it).

<br/>
<details>
  <summary><b>Need a hint? Click here</b> </summary>
  You can use task 3 to figure out which state you are given.
</details>

### Solution

Explaination

In [4]:
%kata T04_CorrectErrorOnLeftQubit

operation CorrectErrorOnLeftQubit (register : Qubit[]) : Unit {
    if (Measure([PauliZ, PauliZ], register[0 .. 1]) == One) {
        X(register[0]);
    }
}

Success!

[Return to Task 4. of the QEC Bit Flip Code Kata.](./QEC_BitFlipCode.ipynb#Task-4.-Error-Correction-I)

### Task 5. Error Detection II

**Input:** three qubits that are either in the state $|\bar{\psi}\rangle = \alpha |000\rangle + \beta |111\rangle$ or in one of the states $X\mathbb{11} |\bar{\psi}\rangle$, $\mathbb{1}X\mathbb{1} |\bar{\psi}\rangle$ or $\mathbb{11}X |\bar{\psi}\rangle$ (i.e., state $|\bar{\psi}\rangle$ with an X error applied to one of the qubits).

**Goal:** determine whether an X error has occurred, and if so, on which qubit.

| Error                     | Output |
|---------------------------|--------|
| None                      | 0      |
| $X\mathbb{11}$            | 1      |
| $\mathbb{1}X\mathbb{1}$   | 2      |
| $\mathbb{11}X$            | 3      |

After applying the operation the state of the qubits should not change.

### Solution

Explaination

In [5]:
%kata T05_DetectErrorOnAnyQubit

operation DetectErrorOnAnyQubit (register : Qubit[]) : Int {
    let m1 = Measure([PauliZ, PauliZ], register[0 .. 1]);
    let m2 = Measure([PauliZ, PauliZ], register[1 .. 2]);
    
    if (m1 == One and m2 == Zero) {
        return 1;
    }
    
    if (m1 == One and m2 == One) {
        return 2;
    }
    
    if (m1 == Zero and m2 == One) {
        return 3;
    }
    
    return 0;
}

Testing with no error.
Testing with error on qubit 1.
Testing with error on qubit 2.
Testing with error on qubit 3.


Success!

[Return to Task 5. of the QEC Bit Flip Code Kata.](./QEC_BitFlipCode.ipynb#Task-5.-Error-Detection-II)

### Task 6. Error Correction II

**Input:** three qubits that are either in the state $|\bar{\psi}\rangle = \alpha |000\rangle + \beta |111\rangle$ or in one of the states $X\mathbb{11} |\bar{\psi}\rangle$, $\mathbb{1}X\mathbb{1} |\bar{\psi}\rangle$ or $\mathbb{11}X |\bar{\psi}\rangle$ (i.e., the qubits start in state $|\bar{\psi}\rangle$ with an X error possibly applied to one of the qubits).

**Goal:**  make sure that the qubits are returned to the state $|\bar{\psi}\rangle$ (i.e., determine whether an X error has occurred on any qubit, and if so, fix it).

### Solution

Explaination

In [6]:
%kata T06_CorrectErrorOnAnyQubit

operation CorrectErrorOnAnyQubit (register : Qubit[]) : Unit {
    let idx = DetectErrorOnAnyQubit(register);
    if (idx > 0) {
        X(register[idx - 1]);
    }
}

Task 06: Testing on [PauliI,PauliI,PauliI]...
Task 06: Testing on [PauliX,PauliI,PauliI]...
Task 06: Testing on [PauliI,PauliX,PauliI]...
Task 06: Testing on [PauliI,PauliI,PauliX]...


Success!

[Return to Task 6. of the QEC Bit Flip Code Kata.](./QEC_BitFlipCode.ipynb#Task-6.-Error-Correction-II)

> All the tasks in this kata have been dealing with X errors on single qubit.  The bit-flip code doesn't allow one to detect or correct a Z error or multiple X errors.  Indeed, 
  * A Z error on a logical state $|\psi\rangle = \alpha |0\rangle + \beta |1\rangle$ encoded using the bit-flip code would convert the encoded state $|\bar{\psi}\rangle = \alpha |000\rangle + \beta |111\rangle$ into $\alpha |000\rangle - \beta |111\rangle$,  which is a correct code word for logical state $\alpha |0\rangle - \beta |1\rangle$.  
  * Two X errors (say, on qubits 1 and 2) would convert $|\bar{\psi}\rangle$ to $\alpha |110\rangle + \beta |001\rangle$,  which is a code word for logical state $\beta |0\rangle + \alpha |1\rangle$ with one X error on qubit 3.

### Task 7. Logical X Gate

**Input:** three qubits that are either in the state $|\bar{\psi}\rangle = \alpha |000\rangle + \beta |111\rangle$ or in one of the states $X\mathbb{11} |\bar{\psi}\rangle$, $\mathbb{1}X\mathbb{1} |\bar{\psi}\rangle$ or $\mathbb{11}X |\bar{\psi}\rangle$ (i.e., state $|\bar{\psi}\rangle$ with an X error applied to one of the qubits).

**Goal:** apply a logical X operator, i.e., convert the qubits to the state $\bar{X} |\bar{\psi}\rangle = \beta |000\rangle + \alpha |111\rangle$ or one of the states that can be represented as $\bar{X} |\bar{\psi}\rangle$ with an X error applied to one of the qubits (for example, $\beta |010\rangle + \alpha |101\rangle$).  If the state has an error, you can fix it, but this is not necessary.

### Solution

Explaination

In [7]:
%kata T07_LogicalX

operation LogicalX (register : Qubit[]) : Unit {
    ApplyToEach(X, register);
}

Task 07: Testing on [PauliI,PauliI,PauliI]...
Task 07: Testing on [PauliX,PauliI,PauliI]...
Task 07: Testing on [PauliI,PauliX,PauliI]...
Task 07: Testing on [PauliI,PauliI,PauliX]...


Success!

[Return to Task 7. of the QEC Bit Flip Code Kata.](./QEC_BitFlipCode.ipynb#Task-7.-Logical-X-Gate)

### Task 8. Logical Z Gate

**Input:** three qubits that are either in the state $|\bar{\psi}\rangle = \alpha |000\rangle + \beta |111\rangle$ or in one of the states $X\mathbb{11} |\bar{\psi}\rangle$, $\mathbb{1}X\mathbb{1} |\bar{\psi}\rangle$ or $\mathbb{11}X |\bar{\psi}\rangle$ (i.e., state $|\bar{\psi}\rangle$ with an X error applied to one of the qubits).

**Goal:** apply a logical Z operator, i.e., convert the qubits to the state $\bar{Z} \bar{\psi}\rangle = \alpha |000\rangle - \beta |111\rangle$ or one of the states that can be represented as $\bar{Z} |\bar{\psi}\rangle$ with an X error applied to one of the qubits (for example, $\alpha |010\rangle - \beta |101\rangle$ ). If the state has an error, you can fix it, but this is not necessary.

### Solution

Explaination

In [8]:
%kata T08_LogicalZ

operation LogicalZ (register : Qubit[]) : Unit {
    ApplyToEach(Z, register);
}

Task 08: Testing on [PauliI,PauliI,PauliI]...
Task 08: Testing on [PauliX,PauliI,PauliI]...
Task 08: Testing on [PauliI,PauliX,PauliI]...
Task 08: Testing on [PauliI,PauliI,PauliX]...


Success!

[Return to Task 8. of the QEC Bit Flip Code Kata.](./QEC_BitFlipCode.ipynb#Task-8.-Logical-Z-Gate)